session.go

```diff
@@ -907,12 +912,92 @@ func (s *session) DropPreparedStmt(stmtID uint32) error {
 	return nil
 }
 
+// StmtTxn wraps kv.Transaction to provide a new kv.Transaction.
+// It holds all statement related modification in the buffer before flush to the txn,
+// so if execute statement meets error, the txn won't be made dirty.
+type StmtTxn struct {
+	kv.Transaction
+
+	buf kv.MemBuffer
+	// TODO: Add binlog related data here.
+	// TODO: DirtyDB need to be statement level too.
+}
+
+// Get overrides the Transaction interface.
+func (st *StmtTxn) Get(k kv.Key) ([]byte, error) {
+	val, err := st.buf.Get(k)
+	if kv.IsErrNotFound(err) {
+		val, err = st.Transaction.Get(k)
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(val) == 0 {
+		return nil, kv.ErrNotExist
+	}
+	return val, nil
+}
+
+// Set overrides the Transaction interface.
+func (st *StmtTxn) Set(k kv.Key, v []byte) error {
+	return st.buf.Set(k, v)
+}
+
+// Delete overrides the Transaction interface.
+func (st *StmtTxn) Delete(k kv.Key) error {
+	return st.buf.Delete(k)
+}
+
+// Seek overrides the Transactioin interface.
+func (st *StmtTxn) Seek(k kv.Key) (kv.Iterator, error) {
+	bufferIt, err := st.buf.Seek(k)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	retrieverIt, err := st.Transaction.Seek(k)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return kv.NewUnionIter(bufferIt, retrieverIt, false)
+}
+
+// SeekReverse overrides the Transactioin interface.
+func (st *StmtTxn) SeekReverse(k kv.Key) (kv.Iterator, error) {
+	bufferIt, err := st.buf.SeekReverse(k)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	retrieverIt, err := st.Transaction.SeekReverse(k)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return kv.NewUnionIter(bufferIt, retrieverIt, true)
+}
+
+// FlushStatement flush data changes by the statement to underlying txn.
+func (st *StmtTxn) FlushStatement(succ bool) error {
+	var err error
+	if succ {
+		err = kv.WalkMemBuffer(st.buf, func(k kv.Key, v []byte) error {
+			if len(v) == 0 {
+				return errors.Trace(st.Transaction.Delete(k))
+			}
+			return errors.Trace(st.Transaction.Set(k, v))
+		})
+	}
+	st.buf.Reset()
+	return errors.Trace(err)
+}
+
 func (s *session) Txn() kv.Transaction {
-	return s.txn
+	if s.Transaction == nil {
+		return nil
+	}
+	return &s.StmtTxn
```

Comment:
It's confusing that `s.Txn()` returns the StmtTxn, but `s.Transaction` returns the underlying transaction.

Reply:
1. I don't want to NewMemBuffer every time a new Txn is created, like this:

```
type StmtTxn struct {
    kv.Transaction
    buf MemBuffer    // New MemBuffer together with kv.Transaction
}
type session struct {
    txn *StmtTxn
}
```

2. I don't want change too much of the old code, mixture bug fix with code refine in one PR.
Maybe you can get here:

```
type session struct {
   txn kv.Transaction
   buf MemBuffer
}

func (s *session) Txn() kv.Transaction {
    if txn == nil {
         return nil
    }
    return &StmtTxn{
         s.txn,
         s.buf,
    }
}
```

3. We'd better keep the StmtTxn struct, which will make many things together:
buffer, binlog, dirtytable and so on...

And the result is what you see now.